[inductor] consolidate common GEMM triton param retrieval

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158987
* #158091
* #157699
* __->__ #158015

\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

coconutruben committed:
[inductor] consolidate common GEMM triton param retrieval

\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben commented on torch/_inductor/kernel/bmm.py:
```diff
@@ -205,22 +200,21 @@ def may_require_contiguous(t, meta_t):
-                **mm_options(config, m, n, k, layout),
+                **kwargs,
             )
+            if e is None:
```
it's how the maybe_append_choice is implemented right now, that it returns None or the error it ran into (because the error is supposed to be non fatal). We can
1. leave as is, and add a comment explaining the behavior
2. make maybe_append_choice return different e.g. (success: bool, error: Optional[Exception])
3. make the determination on the outside i.e. compare the length of choices before/after 
let me know what's preferred here

coconutruben replied:
this logging won't be needed for now, so removing it

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

jansel reviewed (changes requested):
Does this PR introduce semantic changes? 
Or is just a refactor? It would be better to seperate out heuristic changes (test the perf impacts!) from refactoring if it is the former.

coconutruben commented:
> Does this PR introduce semantic changes? Or is just a refactor? It would be better to seperate out heuristic changes (test the perf impacts!) from refactoring if it is the former.

The first version changed some semantics, by adding exhaustive_configs support for templates that were not there before. I added an override for that now, so that it's functionally equivalent to before (e.g. int_mm has no support for exhaustive), and a comment on how to remove the override. Other than that it's only a refactor

coconutruben commented on torch/_inductor/kernel/conv.py:
```diff
@@ -603,7 +595,6 @@ def channels_last_conv():
             sympy_product([x.get_size()[0], *x.get_size()[2:]]),
             out_chan,
             in_chan,
-            **mm_config_kwargs(device_type, _is_large_block_for_cpu),
```
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm_common.py#L184

the original does 2 things
1. for cpu, add the large block callable and scale
2. for everything else, optionally add the dtype size

1. we now override preprocess_mm_configs for CPU to handle that inside the config retrieval
2. conv does not pass through the dtype size here, so that never gets added

so it's functionally equivalent

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben commented:
@pytorchbot merge

henrylhtsang commented:
@pytorchbot revert

henrylhtsang commented:
@pytorchbot revert -m "breaking tests" --no-signal

henrylhtsang commented:
@pytorchbot revert -m "breaking tests" -c ignoredsignal

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D78081314).

coconutruben commented:
@pytorchbot merge

coconutruben commented:
@pytorchbot merge -i

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D78081314](https://our.internmc.facebook.com/intern/diff/D78081314)

[ghstack-poisoned]
